### PR TITLE
replace deprecated set.String in storage/validation

### LINKED ISSUE #117027: Here need continue
https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/pkg/apis/storage/validation/validation.go#L393

### PR DESCRIPTION
**Pull request by charles-chenzz**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
add continue to prevent double insert into set and replace deprecated sets.String to sets.Set[string]
#### Which issue(s) this PR fixes:

Fixes #117027 

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/apis/storage/validation/validation.go [hunk 1]
```diff
@@ -383,14 +383,15 @@ func validateCSINodeDriver(driver storage.CSINodeDriver, driverNamesInSpecs sets
 		allErrs = append(allErrs, field.Duplicate(fldPath.Child("name"), driver.Name))
 	}
 	driverNamesInSpecs.Insert(driver.Name)
-	topoKeys := make(sets.String)
+	topoKeys := sets.New[string]()
 	for _, key := range driver.TopologyKeys {
 		if len(key) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath, key))
```
Comment (wzshiming) [inline pkg/apis/storage/validation/validation.go:389]:
```suggestion
			allErrs = append(allErrs, field.Required(fldPath, key))
			continue
```

File: pkg/apis/storage/validation/validation.go [hunk 2]
```diff
@@ -383,14 +383,16 @@ func validateCSINodeDriver(driver storage.CSINodeDriver, driverNamesInSpecs sets
...
+			continue
 		}
 
 		if topoKeys.Has(key) {
 			allErrs = append(allErrs, field.Duplicate(fldPath, key))
+			continue
```
Comment (RomanBednar) [inline pkg/apis/storage/validation/validation.go:395]: What value does it add? The only effect this would have is that we don't insert a key more than once to `topoKeys` which should not matter because its only purpose is to track duplicates here.

What bug are we trying to solve exactly? The [tracking issue](https://github.com/kubernetes/kubernetes/issues/117027) does not explain it well, but sounds like we want to avoid doing the validation twice - which I don't think will happen with the original code. Do we have an evidence/test to prove this maybe?

File: pkg/apis/storage/validation/validation.go [hunk 3]
```diff
@@ -370,7 +370,7 @@ func validateCSINodeDriverAllocatable(a *storage.VolumeNodeResources, fldPath *f
 }
 
 // validateCSINodeDriver tests if CSINodeDriver has valid entries
-func validateCSINodeDriver(driver storage.CSINodeDriver, driverNamesInSpecs sets.String, fldPath *field.Path,
+func validateCSINodeDriver(driver storage.CSINodeDriver, driverNamesInSpecs sets.Set[string], fldPath *field.Path,
```
Comment (RomanBednar) [inline pkg/apis/storage/validation/validation.go:373]: Updating deprecated code should be in a separate commit.

File: pkg/apis/storage/validation/validation.go (hunk 2 above)
Comment (charles-chenzz) [reply pkg/apis/storage/validation/validation.go:395]: I think the problem we trying to solve here is more like a cleanup, to avoid doing validation twice. and the tracking issue didn't provide more info. and I just use my judgement and found out it seems reasonable for me to add this.

File: pkg/apis/storage/validation/validation.go [hunk 4]
```diff
@@ -383,14 +383,16 @@ func validateCSINodeDriver(driver storage.CSINodeDriver, driverNamesInSpecs sets
 		allErrs = append(allErrs, field.Duplicate(fldPath.Child("name"), driver.Name))
 	}
 	driverNamesInSpecs.Insert(driver.Name)
-	topoKeys := make(sets.String)
+	topoKeys := sets.New[string]()
 	for _, key := range driver.TopologyKeys {
 		if len(key) == 0 {
 			allErrs = append(allErrs, field.Required(fldPath, key))
+			continue
```
[the remark this answers is not in the document]
Comment (charles-chenzz) [reply pkg/apis/storage/validation/validation.go:390]: this part may need to ask @wzshiming . could you take a look at this part?

File: pkg/apis/storage/validation/validation.go (hunk 2 above)
Comment (charles-chenzz) [reply pkg/apis/storage/validation/validation.go:395]: I think in this part will cause valid twice, while validaing CSINode Update. this func will finally call to the ``validateCSINodeDriver()``
https://github.com/kubernetes/kubernetes/blob/2f1db33dd54cbdba52c21c75a53c5387f1cd14ab/pkg/apis/storage/validation/validation.go#L300

Comment (RomanBednar) [reply pkg/apis/storage/validation/validation.go:395]: @charles-chenzz  I've checked the current code with unit test, adding a few prints to see how many times `validateCSINodeDriver()` executes. This verifies we correctly validate once for every driver:


```$ go test -v -run TestCSINodeUpdateValidation
=== RUN   TestCSINodeUpdateValidation
Starting success test case for drivers: [io.kubernetes.storage.csi.driver-1 io.kubernetes.storage.csi.driver-2]
VALIDATING: validateCSINodeDriver

VALIDATING: validateCSINodeDriver

Starting success test case for drivers: [io.kubernetes.storage.csi.driver-1]
VALIDATING: validateCSINodeDriver

Starting success test case for drivers: [io.kubernetes.storage.csi.driver-1 io.kubernetes.storage.csi.driver-2 io.kubernetes.storage.csi.driver-3]
VALIDATING: validateCSINodeDriver

VALIDATING: validateCSINodeDriver

VALIDATING: validateCSINodeDriver

```

If you still believe there is a bug please write a unit test that exposes this bug - it should be failing without the proposed patch, and passing with the patch applied.

Comment (charles-chenzz) [reply pkg/apis/storage/validation/validation.go:395]: I tried to write a unit test that expose it but couldn't. so what I think is wrong tbh. how about we remove continue here and only replace the generic set? can we merge that?

Comment (RomanBednar) [reply pkg/apis/storage/validation/validation.go:395]: @charles-chenzz that sounds good to me, thanks!
